[wip] Fleet permissions

raulcabello opened the pull request:
## Issue: <!-- link the issue or issues this PR resolves here -->
<!-- If your PR depends on changes from another pr link them here and describe why they are needed on your solution section. -->
 
## Problem
<!-- Describe the root cause of the issue you are resolving. This may include what behavior is observed and why it is not desirable. If this is a new feature describe why we need this feature and how it will be used. -->
 
## Solution
<!-- Describe what you changed to fix the issue. Relate your changes back to the original issue / feature and explain why this addresses the issue. -->
 
## Testing
<!-- Note: Confirm if the repro steps in the GitHub issue are valid, if not, please update the issue with accurate repro steps. -->

## Engineering Testing
### Manual Testing
<!-- Describe what manual testing you did (if no testing was done, explain why). -->

### Automated Testing
<!-- Ensure there are unit/integration/validation tests added (if possible); describe what cases they cover and do not cover. -->
* Test types added/modified:
    * Unit
    * Integration (Go Framework)
    * Integration (v2prov Framework)
    * Validation (Go Framework)
    * Other - Explain: _EXPLAIN_
    * None
    * _REMOVE NOT APPLICABLE BULLET POINTS ABOVE_
* If "None" - Reason: _EXPLAIN THE REASON_
  <!-- 
  Non-exhaustive list of reasons:
    - Lack of the framework capable of testing this fix/change
    - Tight deadlines / critical priority to get fix/change in - !ensure GH issue is logged to add tests!
    - No application logic is modified by this change, e.g. refactoring/cosmetic/non-code/test change
    - Tests implemented in another PR elsewhere - !ensure GH PR link is added!
    - Other (explain)
  Note: Outside of the exceptions above, the "existing tests cover the changes" is very unlikely to be an acceptable reason as the existing tests generally don't cover the logic changes implemented by this PR 
  -->
* If "None" - GH Issue/PR: _LINK TO GH ISSUE/PR TO ADD TESTS_

Summary: _TODO_

## QA Testing Considerations
<!-- Highlight areas or (additional) cases that QA should test w.r.t a fresh install as well as the upgrade scenarios -->
 
### Regressions Considerations
<!-- Dedicated section to specifically call out any areas that with higher chance of regressions caused by this change, include estimation of probability of regressions -->
_TODO_

Existing / newly added automated tests that provide evidence there are no regressions:
* _TODO_

MbolotSuse reviewed (changes requested):
Some overall comments:
- I would recommend using the existing labels ([GR](https://github.com/rancher/rancher/blob/12a48d56eb2e577ac386134aca1c90aadacd5069/pkg/controllers/management/auth/globalroles/globalrole_handler.go#L39), [GRB](https://github.com/rancher/rancher/blob/12a48d56eb2e577ac386134aca1c90aadacd5069/pkg/controllers/management/auth/globalroles/globalrolebinding_handler.go#L45)) for GR/GRB ownership. 
- Since labels have a limit on length ([source](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set)) you need to use [wrangler.SafeConcatName](https://github.com/rancher/wrangler/blob/39b1c10e34bff1fa7259d42896aa6e5f1b559a1a/pkg/name/name.go#L56) for the labels and an indexer ([example](https://github.com/rancher/rancher/blob/12a48d56eb2e577ac386134aca1c90aadacd5069/pkg/controllers/management/auth/globalroles/enqueue.go#L176)) to do the mapping from role/clusterRole -> GlobalRoleBinding
- Right now the controllers are entirely GRB based. I would recommend splitting this out more - the GR should update the roles/clusterRoles, and the GRB should update the bindings
-  Some of the integrity checks are a bit minimal - keep in mind that we also need to handle the case where the end user messed with the backing object and we need to correct it. I'd recommend enhancing these to be a bit more rigorous.
- I'd prefer consistency in the package structure of GlobalRoles. Since the other fields aren't in their own package, I don't think these should be. That being said, I'm open to keeping it this way if you can make a good case that we would want to migrate the current fields to use this package structure.

raulcabello commented on pkg/controllers/management/auth/globalroles/fleetpermissions/handler.go:
```diff
@@ -0,0 +1,297 @@
+func (h *Handler) reconcilePermissionsResourceRules(globalRoleBinding *v3.GlobalRoleBinding, globalRole *v3.GlobalRole, fleetWorkspaces []*v3.FleetWorkspace) error {
+	fleetWorkspacePermissionName := fleetWorkspaceClusterRulesPrefix + globalRole.Name
+
+	err := h.reconcileRulesClusterRole(globalRole, globalRoleBinding.Name, fleetWorkspacePermissionName)
```
The idea was to avoid creating the backing `ClusterRoles` until there were at least one `GlobalRoleBinding` to minimize the number of resources created. However, I agree with you now that It'd would be better to do it in the GlobalRole reconciler instead considering the reasons you mentioned. Additionally, this will also make easier the validation in the webhook. Changed in https://github.com/rancher/rancher/pull/44901/commits/f7974c82cfdaed74741e02d383e3e435b6f9a850

raulcabello commented:
> Some overall comments:
> 
> * I would recommend using the existing labels ([GR](https://github.com/rancher/rancher/blob/12a48d56eb2e577ac386134aca1c90aadacd5069/pkg/controllers/management/auth/globalroles/globalrole_handler.go#L39), [GRB](https://github.com/rancher/rancher/blob/12a48d56eb2e577ac386134aca1c90aadacd5069/pkg/controllers/management/auth/globalroles/globalrolebinding_handler.go#L45)) for GR/GRB ownership.
> * Since labels have a limit on length ([source](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set)) you need to use [wrangler.SafeConcatName](https://github.com/rancher/wrangler/blob/39b1c10e34bff1fa7259d42896aa6e5f1b559a1a/pkg/name/name.go#L56) for the labels and an indexer ([example](https://github.com/rancher/rancher/blob/12a48d56eb2e577ac386134aca1c90aadacd5069/pkg/controllers/management/auth/globalroles/enqueue.go#L176)) to do the mapping from role/clusterRole -> GlobalRoleBinding
> * Right now the controllers are entirely GRB based. I would recommend splitting this out more - the GR should update the roles/clusterRoles, and the GRB should update the bindings
> * Some of the integrity checks are a bit minimal - keep in mind that we also need to handle the case where the end user messed with the backing object and we need to correct it. I'd recommend enhancing these to be a bit more rigorous.
> * I'd prefer consistency in the package structure of GlobalRoles. Since the other fields aren't in their own package, I don't think these should be. That being said, I'm open to keeping it this way if you can make a good case that we would want to migrate the current fields to use this package structure.

Thanks for the feedback! I'l mark the PR as draft for now while I addressed your comments. I'll be away for a few days, I'll mark as ready for review again once I make the changes.

raulcabello commented on pkg/controllers/management/auth/globalroles/fleetpermissions/handler.go:
```diff
@@ -0,0 +1,297 @@
+							},
+						},
+						Labels: map[string]string{
+							GRBFleetWorkspaceOwnerLabel: globalRoleBinding.Name,
```
I changed it to use the existing `grOwnerLabel`, and used `wrangler.SafeConcatName() `to make sure we don't reach the length limit.

raulcabello commented on pkg/controllers/management/auth/globalroles/fleetpermissions/handler.go:
```diff
@@ -0,0 +1,297 @@
+}
+
+func (h *Handler) reconcilePermissionsResourceRules(globalRoleBinding *v3.GlobalRoleBinding, globalRole *v3.GlobalRole, fleetWorkspaces []*v3.FleetWorkspace) error {
+	fleetWorkspacePermissionName := fleetWorkspaceClusterRulesPrefix + globalRole.Name
```
I changed it to use  `wrangler.SafeConcatName(globalRole.Name, fleetWorkspaceClusterRulesName)` in order to not reach the limit. Similar to what we do for `GlobalRoles` backing roles

raulcabello commented on pkg/controllers/management/auth/globalroles/fleetpermissions/handler.go:
```diff
@@ -0,0 +1,297 @@
+	}
+
+	for _, fleetWorkspace := range fleetWorkspaces {
+		rbName := subject.Name + "-" + fleetWorkspacePermissionName + "-" + fleetWorkspace.Name
```
done, see https://github.com/rancher/rancher/pull/44901#discussion_r1547493906

raulcabello commented on pkg/controllers/management/auth/globalroles/fleetpermissions/handler.go:
```diff
@@ -0,0 +1,297 @@
+			if err != nil && !apierrors.IsNotFound(err) {
+				returnError = multierror.Append(returnError, err)
+			}
+		} else if !reflect.DeepEqual(rb.Subjects, subjects) {
```
Check added for RoleRef too. Label indicating GR owner will be protected in the webhook (we already do that for other backing roles created for a GlobalRole)

raulcabello commented on pkg/controllers/management/auth/globalroles/fleetpermissions/handler.go:
```diff
@@ -0,0 +1,297 @@
+		return err
+	}
+
+	if !equality.Semantic.DeepEqual(cr.Rules, globalRole.InheritedFleetWorkspacePermissions.ResourceRules) {
```
I don't think we need that as the label will be protected in the webhook (we already do that for other backing roles created for a GlobalRole)

raulcabello replied:
https://github.com/rancher/webhook/pull/348 will prevent users from modifying the label

raulcabello commented on pkg/controllers/management/auth/globalroles/fleetpermissions/handler.go:
```diff
@@ -0,0 +1,297 @@
+
+		return err
+	}
+	if !reflect.DeepEqual(crb.Subjects, subjects) {
```
done for RoleRef and Subjects (see previous comment for protecting the label in the webhook)

raulcabello commented on pkg/controllers/management/auth/globalroles/fleetpermissions/handler.go:
```diff
@@ -0,0 +1,297 @@
+		})
+
+		return err
+	} else if !equality.Semantic.DeepEqual(cr.Rules, rules) {
```
see https://github.com/rancher/rancher/pull/44901#discussion_r1547498448

raulcabello commented on pkg/controllers/management/auth/globalroles/fleetpermissions/handler.go:
I moved the fleet permissions new files to the top-level package. I believe the new package makes the code easier to read, but I removed it for the sake of consistency. However, I'd like to keep the new logic in different files as `globalrole_handler.go` is becoming too big, difficult to read, and difficult to unit tests.

raulcabello commented on pkg/controllers/management/auth/globalroles/enqueue.go:
```diff
@@ -198,6 +202,13 @@ func (g *globalRBACEnqueuer) roleBindingEnqueueGRB(_, _ string, obj runtime.Obje
 		logrus.Errorf("unable to convert object: %[1]v, type: %[1]T to a RoleBinding", obj)
 		return nil, nil
 	}
+	if roleBinding.Labels[fleetpermissions.GRBFleetWorkspaceOwnerLabel] != "" {
```
done, please see the latest commits

raulcabello commented on pkg/controllers/management/auth/globalroles/enqueue.go:
```diff
@@ -239,3 +250,71 @@ func (g *globalRBACEnqueuer) namespaceEnqueueGR(_, _ string, obj runtime.Object)
+}
+
+// clusterRoleEnqueueGRB enqueues GlobalRoleBinding when a generated ClusterRole changes.
+func (g *globalRBACEnqueuer) clusterRoleEnqueueGRB(_, _ string, obj runtime.Object) ([]relatedresource.Key, error) {
```
done, please see the latest commits

raulcabello commented on pkg/controllers/management/auth/globalroles/enqueue.go:
```diff
@@ -239,3 +250,71 @@ func (g *globalRBACEnqueuer) namespaceEnqueueGR(_, _ string, obj runtime.Object)
+}
+
+// clusterRoleBindingEnqueueGRB enqueues GlobalRoleBinding when a generated ClusterRoleBinding changes.
+func (g *globalRBACEnqueuer) clusterRoleBindingEnqueueGRB(_, _ string, obj runtime.Object) ([]relatedresource.Key, error) {
```
done, please see the latest commits

MbolotSuse commented on pkg/controllers/management/auth/globalroles/enqueue_test.go:
```diff
@@ -911,3 +914,173 @@ func Test_namespaceEnqueueGR(t *testing.T) {
+	}
+}
+
+func TestClusterRoleBindingEnqueueGRB(t *testing.T) {
```
There's a couple of test cases that would be good to add:
- Nil obj
- Input object not a clusterRoleBinding
- Error when retrieveing from caches

The test cases also seem to use the old label.

raulcabello replied:
added in the latest commit https://github.com/rancher/rancher/pull/44901/commits/723d79e964a44843ce33dce0e3264ece4ac7a5c8

MbolotSuse commented on pkg/controllers/management/auth/globalroles/enqueue_test.go:
```diff
@@ -911,3 +914,173 @@ func Test_namespaceEnqueueGR(t *testing.T) {
+	}
+}
+
+func TestFleetWorkspaceEnqueueGRB(t *testing.T) {
```
One test case that would be good to add:
- Nil object

MbolotSuse commented on pkg/controllers/management/auth/globalroles/fleetworkspace_binding_handler.go:
```diff
@@ -0,0 +1,187 @@
+
+	subjects := []v1.Subject{rbac.GetGRBSubject(globalRoleBinding)}
+	roleref := v1.RoleRef{
+		APIGroup: "rbac.authorization.k8s.io",
```
Nit:
Here and elsewhere, you could use the `rbacv1.GroupName` constant from the import `rbacv1 "k8s.io/api/rbac/v1"`

raulcabello replied:
added in the latest commit https://github.com/rancher/rancher/pull/44901/commits/723d79e964a44843ce33dce0e3264ece4ac7a5c8

MbolotSuse reviewed (commented):
Some small nits, still neeed to fully work through the handler.

raulcabello commented on pkg/controllers/management/auth/globalroles/enqueue_test.go:
```diff
@@ -911,3 +914,173 @@ func Test_namespaceEnqueueGR(t *testing.T) {
 		})
 	}
 }
+func TestClusterRoleEnqueueGRB(t *testing.T) {
```
added in the latest commit https://github.com/rancher/rancher/pull/44901/commits/723d79e964a44843ce33dce0e3264ece4ac7a5c8

MbolotSuse commented on pkg/controllers/management/auth/globalroles/fleetworkspace_role_handler.go:
```diff
@@ -0,0 +1,162 @@
+		})
+
+		return err
+	} else if !equality.Semantic.DeepEqual(cr.Rules, rules) {
```
Nit: It's possible that this isn't as efficient as it could be. DeepEqual on a slice of struct is going to check every struct and look for equality at several levels (including group/resources/verbs). We know that this is going to be a bit more static (with the same resources/groups each time), so we could write something that takes advantage of this to save a few cpu cycles.

I've marked this as a nit (you can ignore if you don't think it's helpful) since:
- This will be a bit more error prone and you'll need to have tests to cover the various cases
- I'd expect that this won't be much faster in most cases.

raulcabello replied:
There is nothing preventing the user to manually change the resources/groups, so I think we should also check them. Anyway I would prefer using `equality.Semantic.DeepEqual` since it's already been tested.

MbolotSuse replied:
Just to be clear, I wasn't saying that we shouldn't check the resources/groups, more that we could use a targeted check to look at them more efficiently (i.e. check if they are len(1) and that the first entry is the exact item that we expect).

That being said, I can agree with the idea that we should use a known, well-tested method over something custom, so I'll resolve this and we can keep using what we have.

MbolotSuse commented on pkg/controllers/management/auth/globalroles/fleetworkspace_role_handler_test.go:
```diff
@@ -0,0 +1,341 @@
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	tests := map[string]struct {
```
Some test cases that would be good to add:
- We could get the first clusterrole, but not the second (for workspace verbs)
- We couldn't update the first/second cluster role

MbolotSuse replied:
I might be missing something but I don't see the first bullet point (an error other than not found when getting the workspace verbs CR) or the second half of the second bullet.

Overall, I would recommend running the go coverage tool and looking to correct most of the missed red lines (`go test -coverprofile=profile.cov` and `go tool cover -html=profile.cov`). You don't have to cover the "new" functions, but you should get most of the other missed lines, particularly the error cases.

raulcabello replied:
test added in the last commit

MbolotSuse commented on pkg/controllers/management/auth/globalroles/globalrolebinding_handler_test.go:
Would be good to add a test where the fleet workspace permissions was called, but we got an error.

raulcabello replied:
test cases added in https://github.com/rancher/rancher/pull/44901/commits/eebee28deca0e1b9e77dca85baf8eef52597a7a3

raulcabello commented on pkg/controllers/management/auth/globalroles/fleetworkspace_role_handler.go:
```diff
@@ -0,0 +1,162 @@
+		return nil
+	}
+	crName := wrangler.SafeConcatName(globalRole.Name, fleetWorkspaceVerbsName)
+	fleetWorkspaces, err := h.fwCache.List(labels.Everything())
```
Let's just skip the backing `ClusterRole` creation if there are no fleetworkspaces besides local. Keep in mind this is granting permission to just the cluster-wide fleetworkspace resource, so they wouldn't have access to the `fleet-local` workspace namespace. This is an edge case because the `GlobalRole` would need to exist before the workspaces are created. Maybe that might happen when doing a back up and restore? 
It should be fix in this [commit](https://github.com/rancher/rancher/pull/44901/commits/047060c586a486a6f6205fdadd19dd10567ca7b7), please have a look.

raulcabello commented on pkg/controllers/management/auth/globalroles/fleetworkspace_role_handler.go:
```diff
@@ -0,0 +1,162 @@
+
+// ReconcileFleetWorkspacePermissions reconciles backing ClusterRoles created for granting permission to fleet workspaces.
+func (h *fleetWorkspaceRoleHandler) reconcileFleetWorkspacePermissions(globalRole *v3.GlobalRole) error {
+	if globalRole.InheritedFleetWorkspacePermissions.WorkspaceVerbs == nil &&
```
Thanks for pointing this out. I totally missed the case when `inheritedFleetWorkspacePermissions` is removed, but it was present before. In addition to what you mention, I think we should remove the backing `ClusterRole` in the following case:
```
apiVersion: management.cattle.io/v3
kind: GlobalRole
metadata:
  name: example
inheritedFleetWorkspacePermissions:
  workspaceVerbs:
  - "get"
```
is changed to:
```
apiVersion: management.cattle.io/v3
kind: GlobalRole
metadata:
  name: example
```
done in https://github.com/rancher/rancher/pull/44901/commits/b628a62d8dbb9bb666c6c22b43f1564ee891f06f

raulcabello commented on pkg/controllers/management/auth/globalroles/fleetworkspace_binding_handler.go:
```diff
@@ -0,0 +1,187 @@
+			reflect.DeepEqual(rb.RoleRef, roleref) {
+			// undo modifications if rb has changed.
+			rb.Subjects = subjects
+			rb.RoleRef = roleref
```
Since the roleRef field of a binding I think we don't need to check if the roleRef have changed. 

> Updating only when the subjects has changed, delete/create separately if the roleRef has changed
delete/create will need to be done by the user since roleRef is inmutable.

I would remove the roleRef, and just update if the subject has changed. What do you think?

raulcabello replied:
> > I would remove the roleRef, and just update if the subject has changed. What do you think?
> 
> Unfortunately I don't think that this will work. An end user could delete the original binding and create a new one with a different roleRef. This needs to correct such an adjustment.

ahh I see, but this is very unlikely to happen. If an end user deletes the original binding, the reconciliation loop will be triggered and the binding with the right roleRef would be created. The only scenario I see this can occur is if the controller is down and the user deletes and recreate the binding. 

I'll delete/create when any integrity check fails in order to cover this edge case.

raulcabello commented on pkg/controllers/management/auth/globalroles/fleetworkspace_binding_handler_test.go:
```diff
@@ -0,0 +1,520 @@
+	}
+}
+
+func TestReconcileFleetWorkspacePermissionsBindings_errors(t *testing.T) {
```
test cases added in https://github.com/rancher/rancher/pull/44901/commits/eebee28deca0e1b9e77dca85baf8eef52597a7a3

raulcabello commented on pkg/controllers/management/auth/globalroles/enqueue_test.go:
```diff
@@ -911,3 +914,249 @@ func Test_namespaceEnqueueGR(t *testing.T) {
+				return mock
+			},
+			wantKeys:         nil,
+			wantErrorMessage: "unable to get GlobalRole gr for RoleBinding : unexpected error",
```
I disagree with you here. I believe there is value in verifying the message. The test would pass if we encounter a different error by simply checking if the error is nil. We are already using `assert.EqualError` and `assert.ErrorContains` in several places in the Rancher code.

MbolotSuse replied:
> The test would pass if we encounter a different error by simply checking if the error is nil. 

Right, this is intentional. If I were to change the error here to something different (i.e. to add more context) I don't want the test to fail. 

Think about it this way - let's say that I wrote a function that called this function. Would it be acceptable for me to process the string error message to check for a specific error message? For example, checking for a not found by doing a `strings.HasSuffix(err.Error(), "NotFound")`? If not, then we shouldn't be using this in our tests.

> We are already using assert.EqualError and assert.ErrorContains in several places in the Rancher code.

I don't think this is a sufficient justification for the use here. Rancher has it's share of issues, we need to evaluate it if solutions fit before copying them into net new code.

raulcabello replied:
> Right, this is intentional. If I were to change the error here to something different (i.e. to add more context) I don't want the test to fail.

I think it is absolutely fine if the test fails when the error message has changed. The test will clearly show it failed because of the error message, and will show the diff. It should be straightforward to update the test with the new message.

> Think about it this way - let's say that I wrote a function that called this function. Would it be acceptable for me to process the string error message to check for a specific error message? For example, checking for a not found by doing a strings.HasSuffix(err.Error(), "NotFound")? If not, then we shouldn't be using this in our tests.

In that case I would create a custom error type and use type assertion (something like [this](https://www.digitalocean.com/community/tutorials/creating-custom-errors-in-go#type-assertions-and-custom-errors)). This would make unit testing easier, but I think it is an overhead we don't really need if we rely on the error messages. We would end up with a lot of custom error types. I'm open to considering this possibility.

> I don't think this is a sufficient justification for the use here. Rancher has it's share of issues, we need to evaluate it if solutions fit before copying them into net new code.

I agree this is not is a sufficient justification for using it. However, I strongly believe it is better to check the error message than just checking if the error is not nil.  To give an example of current tests in Rancher:
- [here](https://github.com/rancher/rancher/blob/v2.8.3/pkg/controllers/management/auth/globalroles/globalrolebinding_handler_test.go#L414-L415) we just check if we want an error. However, we don't specified which error we are expecting. That means the wrong error might be returned, and the test would pass.
- [here](https://github.com/rancher/rancher/blob/v2.9.0-alpha1/pkg/controllers/dashboard/plugin/fscache_test.go#L237)  we check the error message. That means we verify the right error is returned

MbolotSuse replied:
> In that case I would create a custom error type and use type assertion (something like [this](https://www.digitalocean.com/community/tutorials/creating-custom-errors-in-go#type-assertions-and-custom-errors)).

This is a part of my argument here. If we aren't expecting people to use the string error message for meaningful information, then why are we asserting on it in a test? To put it another way, the string error message isn't a part of the interface - the caller has no guarantee of what (if anything) will be included in that part of the error message.

> However, I strongly believe it is better to check the error message than just checking if the error is not nil.

I disagree with you here.

> - [here](https://github.com/rancher/rancher/blob/v2.8.3/pkg/controllers/management/auth/globalroles/globalrolebinding_handler_test.go#L414-L415) we just check if we want an error. However, we don't specified which error we are expecting. That means the wrong error might be returned, and the test would pass.
> - [here](https://github.com/rancher/rancher/blob/v2.9.0-alpha1/pkg/controllers/dashboard/plugin/fscache_test.go#L237) we check the error message. That means we verify the right error is returned

My biggest problem with these comments is that the "right error" isn't based in the code. The interface of the function provides no guarantees on the type of error returned. "Some error" is equally as valid as the current things that you are asserting on. As you have said above, you don't consider this output consumable for a real function, so it really shouldn't be something that you test.

raulcabello commented on pkg/controllers/management/auth/globalroles/fleetworkspace_binding_handler_test.go:
```diff
@@ -0,0 +1,520 @@
+	user    = "user"
+)
+
+func TestReconcileFleetWorkspacePermissionsBindings(t *testing.T) {
```
added in the latest commits

raulcabello commented on pkg/controllers/management/auth/globalroles/fleetworkspace_binding_handler_test.go:
I refactored the tests to use a "testState" carrier object, and it looks much better now. Thanks for the suggestion!

MbolotSuse commented on pkg/controllers/management/auth/globalroles/fleetworkspace_binding_handler.go:
```diff
@@ -0,0 +1,195 @@
+	if err != nil {
+		return fmt.Errorf("unable to get globalRole: %w", err)
+	}
+	if globalRole.InheritedFleetWorkspacePermissions.WorkspaceVerbs == nil &&
+		globalRole.InheritedFleetWorkspacePermissions.ResourceRules == nil {
```
Since each of the individual side functions checks if there respective part is nil, I don't think we get too much value here. In addition, I think this causes problems in the case that we discussed on the role side - if we go from some permissions -> none this will leave bindings around, which isn't good.

raulcabello replied:
that's been removed as part of adding the deletion for the bindings

MbolotSuse commented on pkg/controllers/management/auth/globalroles/fleetworkspace_binding_handler.go:
```diff
@@ -0,0 +1,195 @@
+		desiredRB := backingRoleBinding(grb, gr, fleetWorkspace.Name)
+		if apierrors.IsNotFound(err) {
+			_, err = h.rbClient.Create(desiredRB)
+			if err != nil && !apierrors.IsNotFound(err) {
```
I'm not really sure that a create call can produce a not found error. Can you elaborate on why this is here?

raulcabello replied:
removed

MbolotSuse commented on pkg/controllers/management/auth/globalroles/fleetworkspace_binding_handler.go:
```diff
@@ -0,0 +1,195 @@
+			!reflect.DeepEqual(rb.RoleRef, desiredRB.RoleRef) {
+			// undo modifications if rb has changed.
+			err := h.rbClient.Delete(rb.Namespace, rb.Name, &metav1.DeleteOptions{})
+			if err != nil {
```
I think that we want to check if the error is not found here as well. My thinking is that if we call delete and the binding wasn't found, then that's ok and we can continue on to the create call below.

raulcabello replied:
added ` !apierrors.IsNotFound(err)` to if condition

MbolotSuse commented on pkg/controllers/management/auth/globalroles/fleetworkspace_binding_handler.go:
```diff
@@ -0,0 +1,195 @@
+		!reflect.DeepEqual(crb.RoleRef, desiredCRB.RoleRef) {
+		// undo modifications if crb has changed.
+		err := h.crbClient.Delete(crb.Name, &metav1.DeleteOptions{})
+		if err != nil {
```
Similar comment here about checking for a not found error.

raulcabello replied:
done

MbolotSuse commented on pkg/controllers/management/auth/globalroles/fleetworkspace_role_handler.go:
```diff
@@ -0,0 +1,191 @@
+func (h *fleetWorkspaceRoleHandler) reconcileResourceRules(gr *v3.GlobalRole) error {
+	crName := wrangler.SafeConcatName(gr.Name, fleetWorkspaceClusterRulesName)
+	cr, err := h.crCache.Get(crName)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return err
+	}
+
+	if gr.InheritedFleetWorkspacePermissions.ResourceRules == nil {
+		return h.deleteClusterRoleIfNeeded(!apierrors.IsNotFound(err), crName)
+	}
+
+	if apierrors.IsNotFound(err) {
+		_, err := h.crClient.Create(backingResourceRulesClusterRole(gr, crName))
+
+		return err
+	}
```
Nit: We check if the error is a not found error 3 times in this section of code.  You could re-write this to minimize the amount of checking that we do:

```suggestion
	if err != nil {
	        if !apierrors.IsNotFound(err){
		        return err
	        }
	        if gr.InheritedFleetWorkspacePermissions.ResourceRules == nil {
	               return h.crClient.Delete(crName, &metav1.DeleteOptions{})
	        }
	        _, err := h.crClient.Create(backingResourceRulesClusterRole(gr, crName))
		return err
	} 
```

raulcabello replied:
This suggestion code is not correct. This would always call `Delete` for `GlobalRoles` that doesn't have `InheritedFleetWorkspacePermissions`.

I refactored the code to minimize the not found check.

MbolotSuse commented on pkg/controllers/management/auth/globalroles/fleetworkspace_role_handler.go:
```diff
@@ -0,0 +1,191 @@
+	}
+}
+
+func (h *fleetWorkspaceRoleHandler) deleteClusterRoleIfNeeded(crExists bool, crName string) error {
```
Nit: I don't think that we get a lot of value out of this side function. The owning function can determine when is appropriate to delete, and since the act of deleting is a one-line call, I don't think is saves us much.

raulcabello replied:
removed

MbolotSuse commented on pkg/controllers/management/auth/globalroles/fleetworkspace_role_handler.go:
```diff
@@ -0,0 +1,191 @@
+func (h *fleetWorkspaceRoleHandler) reconcileWorkspaceVerbs(gr *v3.GlobalRole) error {
+	crName := wrangler.SafeConcatName(gr.Name, fleetWorkspaceVerbsName)
+	cr, err := h.crCache.Get(crName)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return err
+	}
+	crMissing := apierrors.IsNotFound(err)
+
+	if gr.InheritedFleetWorkspacePermissions.WorkspaceVerbs == nil {
+		return h.deleteClusterRoleIfNeeded(!apierrors.IsNotFound(err), crName)
+	}
```
Similar to the above, we check if the error is not found 3 times here. I would recommend re-writing this to minimize the amount of checking we do for this value.

raulcabello replied:
I refactored a bit this part too. However, in this case we have to check more times the `isNotFound` error. That's why I added the `crMissing` var. The reason for that is that I want to prevent listing all fleetworkspaces if `InheritedFleetWorkspacePermissions.WorkspaceVerbs` is nil, which will happens for all existing `GlobalRoles` that doesn't have `InheritedFleetWorkspacePermissions`

MbolotSuse reviewed (changes requested):
Some overall issues:
- While the bug of "some permissions" -> "no permissions" seems to be fixed on the role side, it seems to still be there on the binding side. We want to fix it in both cases - having a binding that points to a non-existent role could cause a security problem.
- There's some missed test cases. I would recommend running tests with `go test -coverprofile=profile.cov` and `go tool cover -html=profile.cov`. You don't have to fix the tests which cover constructors (like `newFleetWorkspaceBindingHandler`) but you should cover all of the branches in the sub-functions.
- The bindings handler should probably check for a `notFound` error when deleting. In these cases the thing that we are trying to delete is already gone, so we are ok to proceed to attempting to create the binding. That being said, I think the current route is ok since the error should cause the handler to be re-enqueued, which should then do the create based on the fact that the binding doesn't exist. 
- Some nit's on code structure (there's a few places where we check if a single error is a not found error 3 times).
- Unresolved thread on checking the error string in tests (we can continue to discuss in thread, just noting it here).

MbolotSuse commented on pkg/controllers/management/auth/globalroles/fleetworkspace_binding_handler.go:
```diff
@@ -0,0 +1,203 @@
+	}
+
+	if gr.InheritedFleetWorkspacePermissions.ResourceRules == nil {
+		return h.crbClient.Delete(crbName, &metav1.DeleteOptions{})
```
Like in other cases, I think that we want to check for a not found error here. Otherwise, this could cause the controller to infinitely loop if the CRB is deleted before this function is called.

MbolotSuse reviewed (changes requested):
A couple of small issues:
- A few missed test cases.
- A few small bugs.
- A few suggestions on resiliency/edge case security.

raulcabello commented on pkg/controllers/management/auth/globalroles/fleetworkspace_binding_handler.go:
```diff
@@ -0,0 +1,203 @@
+	}
+
+	if gr.InheritedFleetWorkspacePermissions.ResourceRules == nil {
+		return h.crbClient.Delete(crbName, &metav1.DeleteOptions{})
```
done

raulcabello commented on pkg/controllers/management/auth/globalroles/fleetworkspace_binding_handler.go:
```diff
@@ -0,0 +1,203 @@
+			// undo modifications if rb has changed.
+			err := h.rbClient.Delete(rb.Namespace, rb.Name, &metav1.DeleteOptions{})
+			if err != nil && !apierrors.IsNotFound(err) {
+				returnError = multierror.Append(returnError, err)
```
I added a continue after this, so it'll skip creation if deletion fails. Anyway it would have retried in the next reconciliation loop because the err of deletion and the possible creation err would've been returned

raulcabello commented on pkg/controllers/management/auth/globalroles/fleetworkspace_role_handler.go:
```diff
@@ -0,0 +1,185 @@
+	if err != nil && !apierrors.IsNotFound(err) {
+		return err
+	}
+	crMissing := apierrors.IsNotFound(err)
```
done

raulcabello commented on pkg/controllers/management/auth/globalroles/fleetworkspace_role_handler.go:
```diff
@@ -0,0 +1,185 @@
+}
+
+// ReconcileFleetWorkspacePermissions reconciles backing ClusterRoles created for granting permission to fleet workspaces.
+func (h *fleetWorkspaceRoleHandler) reconcileFleetWorkspacePermissions(gr *v3.GlobalRole) error {
```
done in the last commit. I have also changed the way error are unit tested (I'll comment in the open thread)

raulcabello commented on pkg/controllers/management/auth/globalroles/fleetworkspace_binding_handler.go:
```diff
@@ -0,0 +1,203 @@
+				continue
+			}
+			if gr.InheritedFleetWorkspacePermissions.ResourceRules != nil {
+				_, err = h.rbClient.Create(desiredRB)
```
I added a check to verify the `ClusterRole` exists. It'll return an error and re-enqueue if it doesn't exist.

raulcabello commented on pkg/controllers/management/auth/globalroles/fleetworkspace_binding_handler.go:
```diff
@@ -0,0 +1,203 @@
+			return err
+		}
+		if gr.InheritedFleetWorkspacePermissions.ResourceRules != nil {
+			_, err = h.crbClient.Create(desiredCRB)
```
I added a check to verify the `ClusterRole` exists. It'll return an error and re-enqueue if it doesn't exist.

raulcabello commented on pkg/controllers/management/auth/globalroles/fleetworkspace_role_handler.go:
```diff
@@ -0,0 +1,185 @@
+	if err != nil {
+		return err
+	}
+	if len(workspacesNames) == 0 {
```
This check is here in the very unlikely event this is called before workspaces are created. 

> If I go from 1 workspace -> 0 workspaces

This should never happens. It should always exist at least one workspace (`fleet-default`). 
If you delete `fleet-default`, you can't import clusters. I see the following error when trying to import a cluster:
```
namespaces "fleet-default" not found
```
`fleet-default` workspace will be recreated when rancher restarts. I think this should be fixed in the webhook by preventing the deletion of the `fleet-default` workspace.

MbolotSuse replied:
This makes sense to me. If you can open up another issue for the fleet-default part I would appreciate it.

raulcabello commented on pkg/controllers/management/auth/globalroles/fleetworkspace_binding_handler_test.go:
```diff
@@ -0,0 +1,307 @@
+		rbCache   *fake.MockCacheInterface[*rbac.RoleBinding]
+		fwCache   *fake.MockNonNamespacedCacheInterface[*v3.FleetWorkspace]
+	}
+	tests := map[string]struct {
```
test cases added. Code coverage for new files is 98.6% and 98.5% (everything except new method)

raulcabello commented on pkg/controllers/management/auth/globalroles/fleetworkspace_binding_handler_test.go:
```diff
@@ -0,0 +1,307 @@
+					ObjectMeta: metav1.ObjectMeta{Name: wrangler.SafeConcatName(grb.Name, fleetWorkspaceVerbsName)},
+				}, nil)
+				state.crbClient.EXPECT().Delete(wrangler.SafeConcatName(grb.Name, fleetWorkspaceVerbsName), &metav1.DeleteOptions{}).Return(errors.NewServiceUnavailable("unexpected error"))
+				state.rbClient.EXPECT().Create(backingClusterRoleBinding(grb, grVerbs, wrangler.SafeConcatName(grb.Name, fleetWorkspaceVerbsName))).Return(nil, errors.NewServiceUnavailable("unexpected error"))
```
test cases added. Code coverage for new files is 98.6% and 98.5% (everything except new method)

raulcabello commented on pkg/controllers/management/auth/globalroles/fleetworkspace_role_handler_test.go:
```diff
@@ -0,0 +1,356 @@
+		fwCache  *fake.MockNonNamespacedCacheInterface[*v3.FleetWorkspace]
+	}
+
+	tests := map[string]struct {
```
test cases added. Code coverage for new files is 98.6% and 98.5% (everything except new method)

MbolotSuse commented on pkg/controllers/management/auth/globalroles/enqueue.go:
```diff
@@ -239,3 +242,83 @@ func (g *globalRBACEnqueuer) namespaceEnqueueGR(_, _ string, obj runtime.Object)
+
+// fleetWorkspaceEnqueueGRB enqueues GlobalRole that have set InheritedFleetWorkspacePermissions
+// when a FleetWorkspace has changed.
+func (g *globalRBACEnqueuer) fleetWorkspaceEnqueueGR(_, _ string, obj runtime.Object) ([]relatedresource.Key, error) {
```
For the `inheritedClusterRoles` part, we add an annotation so that future calls to this method won't cause a mass-re-enqueue of all matching global roles. You can see that [here](https://github.com/rancher/rancher/blob/6148fd36c961f3d3e3210478b686348d7f6f994f/pkg/controllers/management/auth/globalroles/enqueue.go#L117). While I think that this is important for the cluster use case (since I expect that object to see a lot of changes), I don't think it's strictly as important for this use case. Can you take a look at this and see if you want to add something similar here?

raulcabello replied:
I don't think we need that in this case. `FleetWorkspace`  just contains `metav1.ObjectMeta`, and we just use the name of the workspace to generate the backing namespace. Unlike `Clusters`, no field in a `FleetWorkspace` is expected to change that would trigger a re-enqueue of all matching `GlobalRoles`.

MbolotSuse replied:
Alright, that's ok. Let's keep an eye on this from a performance standpoint, but we don't need to fix this here.

MbolotSuse reviewed (commented):
Mostly LGTM. One new comment about the fleetworkspace changed -> enqueue all GRs that give workspace permissions - that has some performance impact that we want to consider here.

JonCrowther commented on pkg/controllers/management/auth/globalroles/fleetworkspace_binding_handler.go:
```diff
@@ -0,0 +1,225 @@
+
+	wrangler "github.com/rancher/wrangler/v2/pkg/name"
+
+	"github.com/hashicorp/go-multierror"
```
multierror has the MPL2.0 license, which we prefer to avoid where possible. Could you use [errors.Join](https://pkg.go.dev/errors#Join)? I realize we use it in other places, but we should avoid it going forward
Something like this: https://github.com/rancher/webhook/blob/54bba7ed560004f40cd277aef921bdcd07802ea6/pkg/resources/management.cattle.io/v3/globalrolebinding/validator.go#L147

raulcabello replied:
done

JonCrowther commented on pkg/controllers/management/auth/globalroles/fleetworkspace_binding_handler.go:
```diff
@@ -0,0 +1,225 @@
+				continue
+			}
+			if gr.InheritedFleetWorkspacePermissions != nil && gr.InheritedFleetWorkspacePermissions.ResourceRules != nil {
+				_, err = h.crCache.Get(desiredRB.RoleRef.Name)
```
I may misunderstand how the fleetworkspace binding handler works, but should this be the rolebinding cache?
```suggestion
				_, err = h.rbCache.Get(desiredRB.RoleRef.Name)
```

raulcabello replied:
No, this is verifying that the `ClusterRole` that this `RoleBinding` is bound to exists.
